{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#general \n",
    "import os\n",
    "import argparse\n",
    "import pickle\n",
    "import dill\n",
    "\n",
    "#you cannot live without \n",
    "from tqdm import trange\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import time\n",
    "#import matplotlib.pyplot as plt\n",
    "import random\n",
    "from termcolor import colored\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "\n",
    "#pip install transformers\n",
    "#pytorch related\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "\n",
    "#bert related\n",
    "from transformers import BertModel, BertTokenizer, BertForSequenceClassification, BertConfig\n",
    "from transformers import AdamW\n",
    "from transformers import get_linear_schedule_with_warmup\n",
    "\n",
    "#self-defined\n",
    "from dataProcessing_bertExtreme import DataProcess\n",
    "import utils_bertExtreme as ut \n",
    "from clfbert import clfModel\n",
    "from eval_metrics import Metrics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mixed pairs dataframe loaded\n",
      "length of the corpus 9514699\n",
      "sample of the corpus ['Mineralocorticoid Receptor (MR) trans-Activation of Inflammatory AP-1 Signaling: DEPENDENCE ON DNA SEQUENCE, MR CONFORMATION, AND AP-1 FAMILY MEMBER EXPRESSION. Glucocorticoids are commonly used to treat inflammatory disorders. The glucocorticoid receptor (GR) can tether to inflammatory transcription factor complexes, such as NFκB and AP-1, and trans-repress the transcription of cytokines, chemokines, and adhesion molecules. In contrast, aldosterone and the mineralocorticoid receptor (MR) primarily promote cardiovascular inflammation by incompletely understood mechanisms. Although MR has been shown to weakly repress NFκB, its role in modulating AP-1 has not been established. Here, the effects of GR and MR on NFκB and AP-1 signaling were directly compared using a variety of ligands, two different AP-1 consensus sequences, GR and MR DNA-binding domain mutants, and siRNA knockdown or overexpression of core AP-1 family members. Both GR and MR repressed an NFκB reporter without influencing p65 or p50 binding to DNA. Likewise, neither GR nor MR affected AP-1 binding, but repression or activation of AP-1 reporters occurred in a ligand-, AP-1 consensus sequence-, and AP-1 family member-specific manner. Notably, aldosterone interactions with both GR and MR demonstrated a potential to activate AP-1. DNA-binding domain mutations that eliminated the ability of GR and MR to cis-activate a hormone response element-driven reporter variably affected the strength and polarity of these responses. Importantly, MR modulation of NFκB and AP-1 signaling was consistent with a trans-mechanism, and AP-1 effects were confirmed for specific gene targets in primary human cells. Steroid nuclear receptor trans-effects on inflammatory signaling are context-dependent and influenced by nuclear receptor conformation, DNA sequence, and the expression of heterologous binding partners. Aldosterone activation of AP-1 may contribute to its proinflammatory effects in the vasculature.', 'RacGAP1 Is a Novel Downstream Effector of E2F7-Dependent Resistance to Doxorubicin and Is Prognostic for Overall Survival in Squamous Cell Carcinoma. We have previously shown that E2F7 contributes to drug resistance in head and neck squamous cell carcinoma (HNSCC) cells. Considering that dysregulation of responses to chemotherapy-induced cytotoxicity is one of the major reasons for treatment failure in HNSCC, identifying the downstream effectors that regulate E2F7-dependent sensitivity to chemotherapeutic agents may have direct clinical impact. We used transcriptomic profiling to identify candidate pathways that contribute to E2F7-dependent resistance to doxorubicin. We then manipulated the expression of the candidate pathway using overexpression and knockdown in in vitro and in vivo models of SCC to demonstrate causality. In addition, we examined the expression of E2F7 and RacGAP1 in a custom tissue microarray (TMA) generated from HNSCC patient samples. Transcriptomic profiling identified RacGAP1 as a potential mediator of E2F7-dependent drug resistance. We validated E2F7-dependent upregulation of RacGAP1 in doxorubicin-insensitive SCC25 cells. Extending this, we found that selective upregulation of RacGAP1 induced doxorubicin resistance in previously sensitive KJDSV40. Similarly, stable knockdown of RacGAP1 in insensitive SCC25 cells induced sensitivity to doxorubicin in vitro and in vivo. RacGAP1 expression was validated in a TMA, and we showed that HNSCCs that overexpress RacGAP1 are associated with a poorer patient overall survival. Furthermore, E2F7-induced doxorubicin resistance was mediated via RacGAP1-dependent activation of AKT. Finally, we show that SCC cells deficient in RacGAP1 grow slower and are sensitized to the cytotoxic actions of doxorubicin in vivo. These findings identify RacGAP1 overexpression as a novel prognostic marker of survival and a potential target to sensitize SCC to doxorubicin.']\n",
      "length of the corpus 9516919\n",
      "sample of the corpus ['Su(H) bound regions in Su(H) induced wing-disc hyperplasia The outcome of Notch activation on proliferation depends on cellular context. In Drosophila wing discs Notch activation causes hyperplasia despite having localized inhibitory effects on proliferation. To understand the underlying mechanisms we have used genomic strategies to identify the Notch-Su(H) target genes directly activated during wing disc hyperplasia. These data are the results from ChIP-Chip experiments to identify genomic regions occupied by Su(H) in hyperplastic Su(H)-expressing Drosophila wing discs.', 'Human subcutaneous adipose tissue gene expression (SOS Sib-pair study, offspring cohort) Obesity has considerable effects on morbidity and mortality, and the prevalence of obesity has been increasing rapidly worldwide during the past two decades. Even if obesity affects the entire individual, adipose tissue plays a central role in the development of obesity. Expression profiling of adipose tissue may give insights into the mechanisms contributing to obesity and obesity-related disorders.The Swedish Obese Subjects (SOS) Sib-Pair Study consists of 154 nuclear families with BMI-discordant sib pairs (BMI difference more than 10 kg/m2) resulting in a study population consisting of 732 subjects. The full SOS Sib-Pair study offspring cohort consists of 425 subjects. Microarray expression analysis in subcutaneous adipose tissue was performed in 375 subjects (262 women and 113 men) of the SOS Sib-Pair offspring cohort.']\n"
     ]
    }
   ],
   "source": [
    "def main():  \n",
    "        \n",
    "    #for calling the file from terminal \n",
    "    parser = argparse.ArgumentParser(description = 'BERT model for data to paper recommendation')\n",
    "    #do aruguments here when not calling from terminal/inside jupyter notebook \n",
    "    args = parser.parse_args([])\n",
    "    args.data_path1 = 'data/'\n",
    "    args.data_path2 = 'IIdata/'\n",
    "    args.subpath = 'sensitivity1vs5000_gene/'\n",
    "    args.load_pretrained = False\n",
    "    args.load_path= 'model_save_v5_sensitivity1vs5000_gene/'\n",
    "    #args.tf_path = '../biobert_v1.1_pubmed'\n",
    "    #args.from_tf= False\n",
    "    args.split1 = True\n",
    "    args.newSplit1= True\n",
    "    args.split2 = True\n",
    "    args.newSplit2 = True \n",
    "\n",
    "    \n",
    "    args.cuda_device = 1\n",
    "    args.learning_rate = 2e-5\n",
    "    args.epsilon = 1e-8\n",
    "    args.train_epochs = 4\n",
    "    args.plot_train = True\n",
    "    args.names1 = []\n",
    "    args.names2 = ['geo','srastudies'] #'immport', 'imspace', 'itnshare',\n",
    "    args.train_ratio = 5000\n",
    "    args.extra = True\n",
    "    \n",
    "    #make sure results are replicable\n",
    "    seed_val = 1234\n",
    "    ut.set_seed(seed_val)\n",
    "    \n",
    "    #load dataloader\n",
    "    dp2 =  DataProcess(path= args.data_path2,\n",
    "                       subpath = args.subpath,\n",
    "          load_pretrained = args.load_pretrained, \n",
    "          load_path = args.load_path,\n",
    "          split = args.split1,\n",
    "          newSplit = args.newSplit2,\n",
    "          names = args.names2,\n",
    "          train_ratio = args.train_ratio,\n",
    "          extra = args.extra)\n",
    "    dp2.dataframize_()\n",
    "    train_loader,valid_loader, test_loader = dp2.dataloaderize_() #dataloader right here, len of records 9639005\n",
    "    \n",
    "    print(len(train_loader), len(valid_loader), len(test_loader)) #  should be less than (218788, 31365, 29))\n",
    "    #check device\n",
    "    if torch.cuda.is_available():\n",
    "        use_cuda = torch.device('cuda:' + str(args.cuda_device))\n",
    "    else:\n",
    "        use_cuda = torch.device('cpu')\n",
    "        \n",
    "    #load model for bert \n",
    "    model = clfModel(load_pretrained = args.load_pretrained, load_path = args.load_path).model\n",
    "    model.to(use_cuda)\n",
    "    \n",
    "    \n",
    "\n",
    "    #optimizer and scheduler\n",
    "    optimizer = AdamW(model.parameters(),\n",
    "                      lr = args.learning_rate,\n",
    "                      eps = args.epsilon)\n",
    "\n",
    "    # Create the learning rate scheduler.\n",
    "    total_steps = len(train_loader) * args.train_epochs\n",
    "    scheduler = get_linear_schedule_with_warmup(optimizer, \n",
    "                                                num_warmup_steps = 0, # Default value in run_glue.py\n",
    "                                                num_training_steps = total_steps)\n",
    "    \n",
    "    #train and valid \n",
    "    training_stats = ut.train(epochs = args.train_epochs,\n",
    "                                     dp2 = dp2,\n",
    "                                     model = model,\n",
    "                                     train_loader = train_loader, \n",
    "                                     valid_loader = valid_loader, \n",
    "                                     optimizer = optimizer, \n",
    "                                     scheduler = scheduler, \n",
    "                                     use_cuda = use_cuda,\n",
    "                                     args = args)\n",
    "    \n",
    "    #plot\n",
    "    if args.plot_train:\n",
    "        ut.plot_train(training_stats, args.load_path)\n",
    "        \n",
    "    \n",
    "        \n",
    "    #prediction on test\n",
    "    combine_predictions, combine_true_labels = ut.predictions( dp2 = dp2,\n",
    "                                                              model = model, \n",
    "                                                              test_loader = test_loader, \n",
    "                                                              use_cuda = use_cuda, \n",
    "                                                              path = args.load_path)\n",
    "    \n",
    "    citation_df = dp2.extra_df.iloc[dp2.test_idx,:]\n",
    "    similarity_dict, max_leng = ut.create_smilarity_dict(citation_df = citation_df, \n",
    "                                                         combine_predictions = combine_predictions, \n",
    "                                                        save_path = args.load_path)\n",
    "    print(max_leng)\n",
    "    #metrics\n",
    "    print('MRR:')\n",
    "    print(Metrics(dp2.citation, leng = max_leng).calculate_mrr(similarity_dict)) #mrr\n",
    "\n",
    "    print('recall@1, recall@10:')\n",
    "    print(Metrics(dp2.citation, leng = max_leng).calculate_recall_at_k(similarity_dict, 1))\n",
    "    print(Metrics(dp2.citation, leng = max_leng).calculate_recall_at_k(similarity_dict, 10))\n",
    "\n",
    "    print('precision@1, precision@10:')\n",
    "    print(Metrics(dp2.citation,leng = max_leng).calculate_precision_at_k(similarity_dict, 1))        \n",
    "    print(Metrics(dp2.citation,leng = max_leng).calculate_precision_at_k(similarity_dict, 10))\n",
    "\n",
    "    print('MAP:')\n",
    "    print(Metrics(dp2.citation,leng = max_leng).calculate_MAP_at_k(similarity_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
